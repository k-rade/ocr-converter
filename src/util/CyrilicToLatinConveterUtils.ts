const CYRILIC_LETTERS =
  "а б в г д ђ е ж з и ј к л љ м н њ о п р с т ћ у ф х ц ч џ ш А Б В Г Д Ђ Е Ж З И Ј К Л Љ М Н Њ О П Р С Т Ћ У Ф Х Ц Ч Џ Ш".split(
    " "
  );
const LATIN_LETTERS =
  "a b v g d đ e ž z i j k l lj m n nj o p r s t ć u f h c č dž š A B V G D Đ E Ž Z I J K L Lj M N Nj O P R S T Ć U F H C Č Dž Š".split(
    " "
  );

export const convertCyrilicToLatin = (string: string): string => {
  return string
    .split("")
    .map((char: string): string => {
      const characterIndex = CYRILIC_LETTERS.indexOf(char);
      if (!~characterIndex) {
        return char;
      }

      return LATIN_LETTERS[characterIndex];
    })
    .join("");
};
