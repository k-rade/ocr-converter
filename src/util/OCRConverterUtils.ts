import Tesseract, { RecognizeResult } from "tesseract.js";
import {
  convertImageError,
  convertImageInit,
  convertImageSuccess,
  setProgress,
} from "../actions/conversionActions";
import { Action } from "../types/Action";
import { Language } from "../types/Language";

export const downloadTextFile = (text: string) => {
  const element = document.createElement("a");
  const file = new Blob([text], {
    type: "text/plain",
  });
  element.href = URL.createObjectURL(file);
  element.download = "myFile.txt";
  element.click();
};

export const copyText = (text: string): void => {
  navigator.clipboard.writeText(text);
};

export const convertImageToText = async (
  dispatch: React.Dispatch<Action>,
  image: string,
  language: string
): Promise<void> => {
  dispatch(convertImageInit());
  let result: RecognizeResult;
  try {
    result = await Tesseract.recognize(image, language, {
      logger: (m) => {
        m.status === "recognizing text" && dispatch(setProgress(m.progress));
      },
    });
    if (result.data.text) {
      dispatch(convertImageSuccess(result.data.text));
    } else {
      dispatch(convertImageError());
    }
  } catch (err) {
    dispatch(convertImageError());
  }
};
