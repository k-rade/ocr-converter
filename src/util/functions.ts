import { toast } from "react-toastify";

export const notify = (message: string, isDarkTheme: boolean) => {
  toast(message, {
    className: isDarkTheme ? "toast-dark" : "toast-light",
    progressClassName: isDarkTheme ? "toast-secondary" : "toast-primary",
  });
};
