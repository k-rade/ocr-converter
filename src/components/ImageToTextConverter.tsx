import { useReducer } from "react";
import OCRFileUpload from "./OCRFileUpload";
import TextWrapper from "./TextWrapper";
import Button from "./Button";
import Languages from "./Languages";
import Error from "./Error";
import { convertImageToText } from "../util/OCRConverterUtils";
import { DEFAULT_LANGUAGE } from "../constants/languages";
import { ErrorType } from "../types/ErrorType";
import { conversionReducer } from "../reducer/conversionReducer";
import {
  initNewConversion,
  setError,
  setImageFile,
  setImageUrl,
  setLanguage,
} from "../actions/conversionActions";

const initialState = {
  language: DEFAULT_LANGUAGE.key,
  imageFile: "",
  imageUrl: "",
  text: "",
  progress: 0,
  isConverting: false,
  error: ErrorType.NO_ERROR,
};

const ImageToTextConverter = (): JSX.Element => {
  const [
    { language, imageFile, imageUrl, text, progress, isConverting, error },
    dispatch,
  ] = useReducer(conversionReducer, initialState);

  const handleImageFileUpload = (value: string): void => {
    dispatch(setImageFile(value));
  };

  const handleLanguageSet = (language: string): void => {
    dispatch(setLanguage(language));
  };

  const handleErrorSet = (error: ErrorType): void => {
    dispatch(setError(error));
  };

  const handleErrorBannerClose = (): void => {
    dispatch(setError(ErrorType.NO_ERROR));
  };

  const handleImageUrlChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(setImageUrl(e.target.value));
  };

  const handleImageToTextConversion = (): void => {
    convertImageToText(dispatch, imageUrl || imageFile, language);
  };

  const handleNewConversion = (): void => {
    dispatch(initNewConversion());
  };

  return (
    <div className="m-auto w-full max-w-lg p-4 pb-14">
      <h1 className="py-5 pb-8 mb-5 font-medium text-3xl text-white w-full border-b border-neutral-800">
        Image to text converter
      </h1>
      <div className="z-10 w-full">
        {error !== ErrorType.NO_ERROR && (
          <Error error={error} onClose={handleErrorBannerClose} />
        )}
        <Languages
          setLanguage={handleLanguageSet}
          selectedLanguage={language}
        />
        <div className="rounded-md w-full mt-5 border border-neutral-800">
          <input
            data-testid="image-url-input"
            type="text"
            placeholder="Enter image URL"
            p-
            className="focus:ring-sky-700 border-0 rounded-t h-20 p-5 text-white outline-0 text-lg placeholder:text-neutral-500 w-full bg-transparent"
            value={imageUrl}
            onChange={handleImageUrlChange}
          />
          <div className="w-full text-center h-0">
            <hr className="w-full -mb-[0.825rem] border-neutral-800" />
            <span className="bg-neutral-900 px-2 text-md text-neutral-500">
              or
            </span>
          </div>
          <OCRFileUpload
            handleError={handleErrorSet}
            setImage={handleImageFileUpload}
            disabled={isConverting}
          />

          <Button
            {...{ disabled: isConverting || !(imageFile || imageUrl) }}
            text="Extract Text"
            fullWidth
            onClick={handleImageToTextConversion}
            isConverting={isConverting}
            conversionProgress={progress}
          />
        </div>
        {text && (
          <TextWrapper text={text} initNewConversion={handleNewConversion} />
        )}
      </div>
    </div>
  );
};

export default ImageToTextConverter;
