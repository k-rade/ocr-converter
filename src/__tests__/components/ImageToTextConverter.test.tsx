import { fireEvent, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Tesseract from "tesseract.js";
import ImageToTextConverter from "../../components/ImageToTextConverter";
import { convertImageToText } from "../../util/OCRConverterUtils";
import { MemoryRouter, useLocation } from "react-router-dom";

jest.mock("tesseract.js", () => ({
  recognize: jest.fn(),
}));

jest.mock("../../util/OCRConverterUtils", () => ({
  convertImageToText: jest.fn(),
}));

jest.mock("react-router-dom", (): any => {
  const actualImplementation = jest.requireActual("react-router-dom");
  return {
    ...actualImplementation,
    useLocation: jest.fn(),
  };
});

describe("ImageToTextConverter", (): void => {
  beforeEach((): void => {
    (useLocation as jest.Mock).mockImplementation(() => {
      return { pathname: "/test-route" };
    });
  });
  const TestComponent = (
    <MemoryRouter>
      <ImageToTextConverter />
    </MemoryRouter>
  );

  it("should render", (): void => {
    const { baseElement } = render(TestComponent);

    expect(baseElement).toMatchSnapshot();
  });

  it("should change image url", (): void => {
    const { getByTestId, getByDisplayValue } = render(TestComponent);

    fireEvent.change(getByTestId("image-url-input"), {
      target: { value: "some-image-url" },
    });

    expect(getByDisplayValue("some-image-url")).toBeInTheDocument();
  });

  it("should upload file", async (): Promise<void> => {
    global.URL.createObjectURL = jest.fn();
    const { getByTestId, getByText } = render(TestComponent);

    const fileInput = getByTestId("ocr-file-upload-input");
    const mockFile = new File(["hello"], "hello.png", { type: "image/png" });

    await waitFor((): void => {
      userEvent.upload(fileInput, mockFile);
      expect(getByText("hello.png")).toBeInTheDocument();
    });
  });

  it("should change language", (): void => {
    const { getByText } = render(TestComponent);

    fireEvent.click(getByText("English"));
    fireEvent.click(getByText("Serbian"));

    expect(getByText("Serbian")).toBeInTheDocument();
  });

  it("should convert image url to text", async (): Promise<void> => {
    const mockResult = {
      data: {
        text: "converted value",
      },
    } as Tesseract.RecognizeResult;
    (Tesseract.recognize as jest.Mock).mockResolvedValue(mockResult);
    global.URL.createObjectURL = jest.fn();

    const { getByTestId, getByText, queryByTestId } = render(TestComponent);

    const fileInput = getByTestId("image-url-input");

    fireEvent.change(fileInput, { target: { value: "some image url" } });

    await waitFor((): void => {
      fireEvent.click(getByText("Convert"));

      expect(convertImageToText).toHaveBeenCalled();
      expect(queryByTestId("error-banner")).not.toBeInTheDocument();
    });
  });

  it("should close error banner", async (): Promise<void> => {
    const { getByText, getByTestId, queryByTestId } = render(TestComponent);

    const fileInput = getByTestId("ocr-file-upload-input");
    const mockFile = new File(["hello"], "hello.txt", { type: "text" });

    await waitFor(() => {
      userEvent.upload(fileInput, mockFile);
    });

    fireEvent.click(getByText("Convert"));

    await waitFor((): void => {
      expect(getByTestId("error-banner")).toBeInTheDocument();
    });

    fireEvent.click(getByText("close.svg"));

    await waitFor((): void => {
      expect(queryByTestId("error-banner")).not.toBeInTheDocument();
    });
  });
});
