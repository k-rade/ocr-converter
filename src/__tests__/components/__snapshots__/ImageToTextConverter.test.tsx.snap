// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageToTextConverter should render 1`] = `
<body>
  <div>
    <div
      class="main-content"
    >
      <div
        class="breadcrumbs"
      >
        <a
          aria-current="page"
          class="breadcrumb breadcrumb-active"
          href="/"
        >
          Home
        </a>
        <span
          class="breadcrumb-divider"
        >
          &gt;
        </span>
        <a
          class="breadcrumb "
          href="/test-route"
        >
          test route
        </a>
      </div>
      <div
        class="ocr-file-upload"
      >
        <input
          class="ocr-file-upload-input"
          data-testid="ocr-file-upload-input"
          type="file"
        />
        <div
          class="ocr-file-upload-container"
        >
          <svg>
            file-upload.svg
          </svg>
          <div
            class="ocr-file-upload-container-txt"
          >
            <h3
              class="ocr-file-upload-container-txt-title"
            >
              Select file to upload
            </h3>
            <p
              class="ocr-file-upload-container-txt-desc"
            >
              Supported formats are: 
              <br />
               jpg, jpeg, png, bmp, pbm. 
              <br />
              Max file upload size is 5 MB.
            </p>
          </div>
        </div>
      </div>
      <div
        class="separator"
      >
        OR 
      </div>
      <div
        class="input-and-languages"
      >
        <input
          class="image-url-input"
          data-testid="image-url-input"
          placeholder="Enter image URL"
          type="text"
          value=""
        />
        <div
          class="languages"
        >
          <div
            class="languages-select"
          >
            <div
              class="selected-language"
            >
              English
              <svg>
                caret-down.svg
              </svg>
            </div>
          </div>
        </div>
      </div>
      <button
        class="btn btn-full-width"
        disabled=""
        type="button"
      >
        Convert
      </button>
      
    </div>
  </div>
</body>
`;
